gamemodes/terrortown/gamemode/server/sv_player_ext.lua

```diff
@@ -1155,57 +1178,62 @@ function plymeta:PickupWeapon(wep, dropBlockingWeapon, shouldAutoSelect)
 		end
 	end
 
-	-- this flag is set to the player to make sure he only picks up this weapon
-	self.wpickup_weapon = wep
+	-- prevent race conditions
+	self.wpickup_waitequip[kind] = true
 
-	-- the flag is set to the weapon to stop other players from auto-picking up this weapon
-	wep.wpickup_player = self
+	timer.Create(kind .. "_WeaponPickup_" .. self:SteamID64(), 0, 1, function()
+		-- this flag is set to the player to make sure he only picks up this weapon
+		self.wpickup_weapon = wep
 
-	-- destroy physics to let weapon float in the air
-	wep:PhysicsDestroy()
+		-- the flag is set to the weapon to stop other players from auto-picking up this weapon
+		wep.wpickup_player = self
 
-	-- set collision group to IN_VEHICLE to be nonexistent, bullets can pass through it
-	wep:SetCollisionGroup(COLLISION_GROUP_IN_VEHICLE)
+		-- destroy physics to let weapon float in the air
+		wep:PhysicsDestroy()
 
-	-- make weapon invisible to prevent stuck weapon in player sight
-	wep:SetNoDraw(true)
+		-- set collision group to IN_VEHICLE to be nonexistent, bullets can pass through it
+		wep:SetCollisionGroup(COLLISION_GROUP_IN_VEHICLE)
 
-	-- set autoselect flag
-	wep.wpickup_autoSelect = shouldAutoSelect
+		-- make weapon invisible to prevent stuck weapon in player sight
+		wep:SetNoDraw(true)
 
-	-- Since we all love GMOD we do some really funny things here. Sometimes the weapon is in
-	-- a position where a player is unable to pick it up, even if there is nothing that hinders
-	-- it from being picked up. Therefore we randomise the position a bit.
-	local function SetWeaponPos()
-		if not IsValid(self) or not IsValid(wep) then return end
+		-- set autoselect flag
+		wep.wpickup_autoSelect = shouldAutoSelect
 
-		-- if a pickup is possible, the weapon gets a flag set and is teleported to the feet
-		-- of the player
-		-- IMPORTANT: If the weapon gets teleported into other entities, it gets stuck. Therefore
-		-- the weapon is teleported to half player height
-		local pWepPos = self:EyePos()
-		pWepPos.z = pWepPos.z - 20 -- -20 to move it outside the viewing area
+		-- Since we all love GMOD we do some really funny things here. Sometimes the weapon is in
+		-- a position where a player is unable to pick it up, even if there is nothing that hinders
+		-- it from being picked up. Therefore we randomise the position a bit.
+		local function SetWeaponPos()
+			if not IsValid(self) or not IsValid(wep) or not self.wpickup_waitequip[kind] then return end
```

Comment:
self should be localized for this localized function...

Reply:
Sure (It's an artifact), but I am curious though. Isn't self just like wep in this context? In my understanding wep is just the second argument to a function while self is the implicit first argument. Since my understanding of the lua stack is very limited I am curious as to how they are different.

Reply:
yea self is a - more or less - localized var in this case (the player's var). But it is just recommended to use self in exact this instance - where it was meant to be defined. Using it outside this situation or inside another function, this is an indicator that there exists inconsistency or something that can be done in a better way. Sometimes, jenkins doesn't like those situation too, so often we use "self" in a enclosed / included other localized function, we declare a var too avoid issues (smth like `local ply = self`).